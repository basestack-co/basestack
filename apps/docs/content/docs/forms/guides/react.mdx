---
title: NextJS & React
icon: Code
tags: Basestack Forms, NextJS, React Integration, Form submission, Fetch API, Axios, React, Next.js, API integration, JSON response, Web development, Developer tools, "2025"
---

import { Step, Steps } from 'fumadocs-ui/components/steps';

[React](https://react.dev/?ref=basestack.co) is a JavaScript library for building user interfaces, allowing developers to create interactive, dynamic web applications with reusable components. [Next.js](https://nextjs.org/?ref=basestack.co) is a React framework that enables server-side rendering, static site generation, and simplified routing for building fast, scalable web applications.

To get started, youâ€™ll need the following:

- A [Basestack Forms](https://forms.basestack.co) Account
- A React or Next.js project with a Form component

<Steps>

<Step>
### Use this on your React Project to send form data to your endpoint.

Send data using the [Fetch API](https://developer.mozilla.org/en-US/docs/Web/API/Fetch_API?ref=basestack.co), [Axios](https://axios-http.com/?ref=basestack.co) or Any [HTTP](https://developer.mozilla.org/en-US/docs/Web/HTTP?ref=basestack.co) Client and ensure the mode on the query string is set to `mode=rest` for retrieving the response in JSON format.
</Step>
<Step>
### Copy and paste this into your React component.

Feel free to use the code below to send a test submission.


```tsx copy title="contact.tsx"
import React, { useState } from "react";

const FormComponent = () => {
  // You can also use rect-hook-form or formik for form handling
  // check out the docs for more info https://react-hook-form.com/
  const [name, setName] = useState("");
  const [email, setEmail] = useState("");
  const [message, setMessage] = useState("");

  const onHandleSubmit = (e) => {
    e.preventDefault();
    e.stopPropagation();

    // You can also use axios or any other library to make the request
    // You can also use async await to make the request
    fetch("https://forms.basestack.co/api/v1/s/[KEY]?mode=rest", {
      method: "POST",
      headers: {
        Accept: "application/json",
        "Content-Type": "application/json",
      },
      body: JSON.stringify({ name, email, message }),
    })
      .then((res) => res.json())
      .then((data) => {
        if (data.code === 200) {
          console.log("Form submitted successfully");
        }
      })
      .catch((error) => {
        console.log("Error submitting form", error);
      });
  };

  return (
    <form onSubmit={(e) => onHandleSubmit(e)}>
      <div>
        <label htmlFor="name">Name:</label>
        {/*  <!-- name each of your inputs as you wish --> */}
        <input
          type="text"
          id="name"
          name="name"
          required
          autoComplete="name"
          aria-label="Name"
          placeholder="Enter your name"
          value={name}
          onChange={(e) => setName(e.target.value)}
        />
      </div>

      <div>
        <label htmlFor="email">Email:</label>
        <input
          type="email"
          id="email"
          name="email"
          required
          autoComplete="email"
          aria-label="Email"
          placeholder="Enter your email"
          value={email}
          onChange={(e) => setEmail(e.target.value)}
        />
      </div>

      <div>
        <label htmlFor="message">Message:</label>
        <textarea
          id="message"
          name="message"
          required
          aria-label="Message"
          placeholder="Enter your message"
          value={message}
          onChange={(e) => setMessage(e.target.value)}
        ></textarea>
      </div>

      {/*  <!-- your other form fields go here --> */}

      <button type="submit">Submit</button>
    </form>
  );
};

export default FormComponent;
```
</Step>
</Steps>
