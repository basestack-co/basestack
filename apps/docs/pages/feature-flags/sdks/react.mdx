---
title: React SDK Reference
description: Integration with React helps and facilitates the process of testing and developing features in production and other environments.
---

import { Steps, Callout } from "nextra/components";

# React SDK Reference

Integration with React helps and facilitates the process of testing and developing features in production and other environments. This integration consists of a set of plug-and-play components to accelerate the development process in your project.

## Getting started

<Steps>
  ### Install

First, let's install some packages!

```sh copy npm2yarn
npm i @basestack/flags-js-sdk @basestack/flags-react-sdk
```

### Create a client provider

In your index.js file, import FlagsProvider from @basestack/flags-react-sdk and include the configuration parameters specific to your Basestack project. You can find these parameter values on your project settings page.

```js copy filename="index.js"
import FlagsJS from "@basestack/flags-js-sdk";
import { FlagsProvider } from "@basestack/flags-react-sdk";

const sdk = new FlagsJS({
  apiUrl: "https://your-basestack-hosted-app-domain.com/api/v1",
  projectKey: "xxxxx",
  envKey: "xxxxx",
});

root.render(
  <FlagsProvider
    sdk={sdk}
    onSuccessfulInit={() => console.log("Successful Init FlagsJS SDK")}
  >
    <App />
  </FlagsProvider>,
);
```

Your app is now fully equipped to take advantage of feature flags and other powerful functionalities offered by Basestack. To maximize the potential of Basestack, refer to the instructions provided in the supported hooks and components documentation.

### Usage

```jsx copy filename="MyComponent.js"
import { useFlag } from "@basestack/flags-react-sdk";

const AppComponent = () => {
  const { enabled } = useFlag("private_msg_2");
  const privateChat = useFlag("private_chat");

  return (
    <div>
      {enabled && <div>This is a great feature</div>}
      {privateChat.enabled && <div>This is a great private chat</div>}
    </div>
  );
};
```

</Steps>

## Cache vs Async

The SDK provides two alternatives for fetching a feature flag: the **useFlag** and **useFlagAsync** hooks. Both serve the same purpose of retrieving a flag by its name. However, the 'useFlag' hook utilizes the cached version, resulting in faster access, while 'useFlagAsync' hook fetches the latest version directly from the server.

Which hook should you choose? It depends on your specific needs. If having the most up-to-date version of the flag is crucial and it changes frequently, then opt for the async hooks. On the other hand, if the flag doesn't undergo frequent changes and you prioritize performance, the hook with cache support are the ideal choice. Assess your requirements and select the method that aligns best with your priorities.

<Callout>
  The cache functions as a state container that automatically updates every time
  your app is restarted or updated.
</Callout>

```js copy filename="main.js"
import { useFlag, useFlagAsync } from "@basestack/flags-react-sdk";

const MyComponent = () => {
  // Get from the cache, no call for the server
  const cache = useFlag("header");
  // Get a new fresh version from the server
  const server = useFlagAsync("header");

  return (
    <div>
      {!server.isLoading && server.flag.enabled && (
        <div>
          This feature is waiting for the server response to check if we can
          enabled
        </div>
      )}
      {cache.enabled && <div>This is enabled feature from the cache</div>}
    </div>
  );
};
```

## Hooks

The library fully supports React hooks, providing the **useFlag**, **useFlagAsync**, **useFlags** and **useFlagsAsync** hook for a programmatic approach.

### useFlag & useFlagAsync Hooks

With **useFlag** or **useFlagAsync**, you can create and utilize multiple flags without any restrictions; simply change the names as needed. For practical implementation, refer to the examples provided.

```js copy
import { useFlag, useFlagAsync } from "@basestack/flags-react-sdk";

const MyComponent = () => {
  const { enabled } = useFlag("private_msg_2");
  const privateChat = useFlag("private_chat");
  // Get a new fresh version from the server
  const server = useFlagAsync("header");

  return (
    <div>
      {!server.isLoading && server.flag.enabled && (
        <div>
          This feature is waiting for the server response to check if we can
          enabled
        </div>
      )}
      {enabled && <div>This is a great feature</div>}
      {privateChat.enabled && <div>This is a great private chat</div>}
    </div>
  );
};
```

#### Remote Flags

Remote Flags share the same characteristics as regular Flags, but they come with an additional advantage – a data payload.

This payload empowers you to furnish real-time properties to your project's features. For instance, if certain features of a project require dynamic color changes, size adjustments, or any other data variations across different environments, utilizing the data payload of a Remote Flag becomes the ideal solution.

```js copy
import { useFlag, useFlagAsync } from "@basestack/flags-react-sdk";

const MyComponent = () => {
  const { enabled, payload } = useFlag("my_remote_flag");
  // Get a new fresh version from the server
  const server = useFlagAsync("header");

  return (
    <div>
      {!server.isLoading && server.flag.enabled && (
        <div style={{ color: server.flag.payload.color }}>
          This feature is waiting for the server response to check if we can
          enabled
        </div>
      )}
      {enabled && (
        <div style={{ color: payload.color }}>
          This is a great remote feature
        </div>
      )}
    </div>
  );
};
```

### useFlags & useFlagsAsync Hooks

For scenarios where you need to fetch all flags simultaneously, make use of the **useFlags** and **useFlagsAsync** hooks. These hooks enable efficient retrieval of multiple flags in a single operation.

```js copy
import { useFlags, useFlagsAsync } from "@basestack/flags-react-sdk";

const MyComponent = () => {
  const allFlags = useFlags();
  // Or use the useFlagsAsync to get fresh list from the server, this will update the cache state
  const allFlagsFromServer = useFlagsAsync();

  return (
    <div>
      {allFlags.map(({ enabled, slug }) => {
        if (!enabled) return null;

        return <p>{slug}</p>;
      })}
    </div>
  );
};
```

## Components

### Flag & FlagAsync Components

The pre-built components **Flag** or **FlagAsync** accepts one or multiple child components. Any components placed inside the **Flag** wrapper will only be rendered when the flag exists or enabled.

```js copy
import { Flag, FlagAsync } from "@basestack/flags-react-sdk";

const AppComponent = () => {
  return (
    <div>
      {/* Get a Flag from the cache */}
      <Flag name="private_msg_2">
        <YourFeature />
      </Flag>

      {/* Get a Flag from the cache and use as a Remote Flag. Ps. You can do the same with the FlagAsync Component */}
      <Flag
        name="my_remote_flag"
        render={(flag) => (
          <div style={{ color: flag.payload.color }}>
            A Remote flag inside a component
          </div>
        )}
      />

      {/* Get a Flag from the server */}
      <FlagAsync name="private_msg_2">
        <YourFeature />
      </FlagAsync>
    </div>
  );
};
```

## TypeScript Support

The SDK is built with TypeScript. Therefore, when utilized in a TypeScript project, it comes with basic type inference out of the box.

```js copy filename="main.ts"
import { useFlag } from "@basestack/flags-react-sdk";

const flag = useFlag("new_header_v2");

flag.enabled; // boolean
flag.createdAt; // Date
flag.slug; // string
flag.description; // string
// ...
```

#### Typescript Type Definition

```sh copy
interface Flag {
  slug: string;
  enabled: boolean;
  payload?: unknown;
  expiredAt?: Date;
  createdAt: Date;
  updatedAt: Date;
  description: string;
}
```

## SDK Reference

## Initialisation options

For details regarding the initialization options of @basestack/flags-js-sdk, please refer to the documentation related to the [Javascript SDK](/feature-flags/sdks/javascript#initialisation-options).

## Troubleshooting

### Why are my components rendering twice?

It's normal. Check if your App/Component is wrapped inside [React.StrictMode ↗](https://react.dev/docs/strict-mode). The purpose of `React.StrictMode` is to help identify potential problems in your application by highlighting some unsafe practices or potential bugs during development.

When you wrap a component in `React.StrictMode`, it can cause some components to be rendered twice. This is expected behavior and is not specific to your code. Rendering twice is intentional and done by React on purpose.

```js copy filename="main.ts"
const root = ReactDOM.createRoot(document.getElementById("root"));

root.render(
  <React.StrictMode>
    <FlagsProvider
      sdk={sdk}
      onSuccessfulInit={() => console.log("Successful Init FlagsJS SDK")}
    >
      <App />
    </FlagsProvider>
  </React.StrictMode>,
);
```

## Examples

**coming soon**
