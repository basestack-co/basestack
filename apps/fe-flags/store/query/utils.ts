import {
  MiddlewareAPI,
  isRejectedWithValue,
  Middleware,
} from "@reduxjs/toolkit";

export const baseUrl = `http://localhost:3000/api/v1`;

export const prepareHeaders = (headers: Headers) => {
  // const state = getState() as RootState;
  return headers;
};

export const rtkQueryErrorLogger: Middleware =
  (api: MiddlewareAPI) => (next) => (action) => {
    if (isRejectedWithValue(action)) {
      const {
        payload: { status },
      } = action;
      // Access forbidden
      if (status === 403) {
        console.log("rtkQueryErrorLogger Access forbidden");
      }
    }
    return next(action);
  };
