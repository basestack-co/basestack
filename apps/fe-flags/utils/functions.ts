// Utils
import isEmpty from "lodash.isempty";

export const isBrowser = typeof window !== "undefined";

export const isNavigator = typeof navigator !== "undefined";

/**
 * UTILS OPERATIONS
 */

export const validURL = (str: string) => {
  const pattern = new RegExp(
    "^(https?:\\/\\/)?" + // protocol
      "((([a-z\\d]([a-z\\d-]*[a-z\\d])*)\\.)+[a-z]{2,}|" + // domain name
      "((\\d{1,3}\\.){3}\\d{1,3}))" + // OR ip (v4) address
      "(\\:\\d+)?(\\/[-a-z\\d%_.~+]*)*" + // port and path
      "(\\?[;&a-z\\d%_.~+=-]*)?" + // query string
      "(\\#[-a-z\\d_]*)?$",
    "i"
  );
  return pattern.test(str);
};

export const urlQueryBuilder = (urlParams: any) => {
  const esc = encodeURIComponent;
  return Object.keys(urlParams)
    .filter((x) => !isEmpty(urlParams[x]))
    .map((k) => esc(k) + "=" + esc(urlParams[k]))
    .join("&");
};

export function groupArrayOfObjects(list: Array<any>, key: string) {
  return list.reduce((rv, x) => {
    (rv[x[key]] = rv[x[key]] || []).push(x);
    return rv;
  }, {});
}

export function groupBy(xs, f) {
  return xs.reduce(
    (r, v, i, a, k = f(v)) => ((r[k] || (r[k] = [])).push(v), r),
    {}
  );
}

export function getBaseUrl() {
  if (typeof window !== "undefined") {
    return "";
  }
  // reference for vercel.com
  if (process.env.VERCEL_URL) {
    return `https://${process.env.VERCEL_URL}`;
  }

  // // reference for render.com
  if (process.env.RENDER_INTERNAL_HOSTNAME) {
    return `http://${process.env.RENDER_INTERNAL_HOSTNAME}:${process.env.PORT}`;
  }

  // assume localhost
  return `http://localhost:${process.env.PORT ?? 3000}`;
}
