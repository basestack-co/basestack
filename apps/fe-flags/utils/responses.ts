// MESSAGES

export const unauthorized = {
  code: 401,
  error: true,
  message: "You must be sign in to view the protected content on this page.",
};

export const methodNotAllowed = {
  code: 405,
  error: true,
  message: `The HTTP method is not supported at this route.`,
};

export const somethingWentWrong = "Something went wrong";

export const notAuthorizedActionProject =
  "You are not authorized to do this action in this project";

// ENVIRONMENTS

export const notAuthorizedCreateEnv =
  "You are not authorized to create an environment";

export const notAuthorizedUpdateEnv =
  "You are not authorized to update this environment";

export const notAuthorizedDeleteEnv =
  "You are not authorized to delete this environment";

// FLAGS

export const notAuthorizedCreateFlag =
  "You are not authorized to create an flag";

export const notAuthorizedUpdateFlag =
  "You are not authorized to update this flag";

export const notAuthorizedDeleteFlag =
  "You are not authorized to delete this flag";

// HISTORY

export const notAuthorizedCreateHistory =
  "You are not authorized to create an history";

export const notAuthorizedUpdateHistory =
  "You are not authorized to update this history";
