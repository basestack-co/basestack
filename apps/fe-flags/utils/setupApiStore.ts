import {
  AnyAction,
  combineReducers,
  configureStore,
  EnhancedStore,
  Middleware,
  Reducer,
} from "@reduxjs/toolkit";

export function setupApiStore<
  A extends {
    reducer: Reducer<any, any>;
    reducerPath: string;
    middleware: Middleware;
    util: { resetApiState(): any };
  },
  R extends Record<string, Reducer<any, any>> = Record<never, never>
>(api: A, extraReducers?: R): { api: any; store: EnhancedStore } {
  /*
   * Modified version of RTK Query's helper function:
   * https://github.com/reduxjs/redux-toolkit/blob/master/packages/toolkit/src/query/tests/helpers.tsx
   */
  const getStore = (): EnhancedStore =>
    configureStore({
      reducer: combineReducers({
        [api.reducerPath]: api.reducer,
        ...extraReducers,
      }),
      middleware: (gdm) =>
        gdm({ serializableCheck: false, immutableCheck: false }).concat(
          api.middleware
        ),
    });

  type StoreType = EnhancedStore<
    {
      api: ReturnType<A["reducer"]>;
    } & {
      [K in keyof R]: ReturnType<R[K]>;
    },
    AnyAction,
    ReturnType<typeof getStore> extends EnhancedStore<any, any, infer M>
      ? M
      : never
  >;

  const initialStore = getStore() as StoreType;
  const refObj = {
    api,
    store: initialStore,
  };
  const store = getStore() as StoreType;
  refObj.store = store;

  return refObj;
}
