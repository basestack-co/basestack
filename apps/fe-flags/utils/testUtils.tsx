import React from "react";
import { render, RenderOptions } from "@testing-library/react";
// Store
import { Provider } from "react-redux";
import { store } from "../store";
// Locales
import { IntlProvider } from "react-intl";
import { messages } from "../locales";
// Styles
import { ThemeProvider } from "styled-components";
import theme from "design-system/theme";
// Auth
import { SessionProvider } from "next-auth/react";

interface AllProvidersProps {
  initialState?: {};
}

const AllProviders: React.FC<AllProvidersProps> = ({
  children,
  initialState = {},
}) => {
  return (
    <SessionProvider session={{ expires: null }}>
      <Provider store={{ ...store, ...initialState }}>
        <IntlProvider locale="en" messages={messages.en}>
          <ThemeProvider theme={theme}>{children}</ThemeProvider>
        </IntlProvider>
      </Provider>
    </SessionProvider>
  );
};

const renderWithAllProviders = (
  ui: React.ReactElement,
  initialState?: {},
  options?: Omit<RenderOptions, "wrapper">
) =>
  render(ui, {
    wrapper: (props) => <AllProviders initialState={initialState} {...props} />,
    ...options,
  });

const WithThemeProvider: React.FC = ({ children }) => (
  <ThemeProvider theme={theme}>{children}</ThemeProvider>
);

const renderWithTheme = (
  ui: React.ReactElement,
  options?: Omit<RenderOptions, "wrapper">
) => render(ui, { wrapper: WithThemeProvider, ...options });

const WithReduxProvider: React.FC = ({ children }) => (
  <Provider store={store}>{children}</Provider>
);

const renderWithRedux = (
  ui: React.ReactElement,
  options?: Omit<RenderOptions, "wrapper">
) => render(ui, { wrapper: WithReduxProvider, ...options });

export { renderWithAllProviders, renderWithTheme, renderWithRedux };
