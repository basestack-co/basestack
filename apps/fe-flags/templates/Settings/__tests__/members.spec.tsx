import React from "react";
import { renderWithAllProviders } from "../../../utils/testUtils";
import { cleanup } from "@testing-library/react";
import Members from "../Members";

describe("Members General tests", () => {
  afterEach(cleanup);

  test("render General correctly", () => {
    const { asFragment } = renderWithAllProviders(<Members />);

    expect(asFragment()).toMatchSnapshot();
  });
});
