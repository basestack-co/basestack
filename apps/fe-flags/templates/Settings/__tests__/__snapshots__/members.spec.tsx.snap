// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Members General tests render General correctly 1`] = `
<DocumentFragment>
  .c12 {
  height: 1.75rem;
  width: 1.75rem;
  border-radius: 50%;
  overflow: hidden;
  -webkit-flex-shrink: 0;
  -ms-flex-negative: 0;
  flex-shrink: 0;
  margin-right: 0.75rem;
  border-radius: 50%;
}

.c13 {
  height: 100%;
  width: 100%;
  object-fit: cover;
}

.c5 {
  font-size: 0.875rem;
  line-height: 1.375rem;
  font-weight: 400;
  margin-bottom: 1.25rem;
  font-family: 'Roboto',sans-serif;
  color: #000000;
}

.c9 {
  font-size: 0.875rem;
  line-height: 1.375rem;
  font-weight: 400;
  font-family: 'Roboto',sans-serif;
  color: #757575;
}

.c14 {
  font-size: 0.875rem;
  line-height: 1.375rem;
  font-weight: 400;
  font-weight: 400;
  font-family: 'Roboto',sans-serif;
  color: #000000;
}

.c21 {
  font-size: 0.875rem;
  line-height: 1.375rem;
  font-weight: 400;
  margin-right: 0.75rem;
  font-family: 'Roboto',sans-serif;
  color: #757575;
}

.c4 {
  font-size: 1.125rem;
  line-height: 1.625rem;
  font-weight: 500;
  margin-bottom: 0.25rem;
  font-family: 'Roboto',sans-serif;
  color: #000000;
}

.c17 {
  height: 1.5rem;
  width: 1.5rem;
  -webkit-flex-shrink: 0;
  -ms-flex-negative: 0;
  flex-shrink: 0;
}

.c18 {
  color: #000000;
  font-size: 1.5rem;
}

.c22 {
  border: none;
  height: 2.25rem;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  padding: 0 0.75rem;
  font-size: 0.875rem;
  font-weight: 500;
  cursor: pointer;
  border-radius: 4px;
  -webkit-text-decoration: none;
  text-decoration: none;
  -webkit-transition: background-color 0.1s ease-in-out,color 0.1s ease-in-out, border-color 0.1s ease-in-out;
  transition: background-color 0.1s ease-in-out,color 0.1s ease-in-out, border-color 0.1s ease-in-out;
  background-color: #000000;
  color: #FFFFFF;
  margin-left: auto;
  width: -webkit-fit-content;
  width: -moz-fit-content;
  width: fit-content;
}

.c22:hover:not(:active) {
  background-color: #333;
}

.c22 .material-symbols-sharp {
  color: inherit;
}

.c2 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  box-shadow: rgba(0,0,0,0.1) 0px 1px 3px 0px,rgba(0,0,0,0.06) 0px 1px 2px 0px;
  background-color: #FFFFFF;
  border-radius: 0.375rem;
}

.c19 {
  height: 1px;
  border: none;
  background-color: #EEEEEE;
  width: 100%;
}

.c16 {
  border: none;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  cursor: pointer;
  border-radius: 50%;
  -webkit-transition: background-color 0.1s ease-in-out,color 0.1s ease-in-out, border-color 0.1s ease-in-out;
  transition: background-color 0.1s ease-in-out,color 0.1s ease-in-out, border-color 0.1s ease-in-out;
  background-color: transparent;
  color: #000000;
  height: 2rem;
  width: 2rem;
}

.c16 .material-symbols-sharp {
  color: inherit;
}

.c16:hover {
  background-color: #EEEEEE;
}

.c16:active {
  background-color: #E2E2E2;
}

.c6 {
  background-color: #F6F6F6;
  padding: 1.25rem;
  border-radius: 4px;
}

.c7 {
  display: grid;
  grid-template-columns: repeat(3,1fr) auto;
  grid-gap: 1.25rem;
}

.c7:not(:last-child) {
  border-bottom: 1px solid #E2E2E2;
}

.c10 {
  min-width: 2.25rem;
  pointer-events: none;
}

.c8 {
  height: 2.8125rem;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
}

.c11 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.c15 {
  height: 2rem;
  width: 2rem;
}

.c3 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  padding: 1.25rem;
}

.c20 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  padding: 1rem 1.25rem;
}

.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
}

.c1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
}

.c1:not(:last-child) {
  margin-bottom: 1.25rem;
}

<ul
    class="c0"
  >
    <ul
      class="c1"
    >
      <div
        class="c2"
        data-testid="card"
      >
        <div
          class="c3"
        >
          <h3
            class="c4"
            data-testid="setting-title"
          >
            Team members
          </h3>
          <p
            class="c5"
            data-testid="setting-title"
          >
            Active Members
          </p>
          <div
            class="c6"
            data-testid="table"
          >
            <div
              class="c7"
              data-testid="header"
            >
              <div
                class="c8"
              >
                <p
                  class="c9"
                >
                  Name
                </p>
              </div>
              <div
                class="c8"
              >
                <p
                  class="c9"
                >
                  Email
                </p>
              </div>
              <div
                class="c8"
              >
                <p
                  class="c9"
                >
                  Role
                </p>
              </div>
              <div
                class="c10"
              />
            </div>
            <div
              class="c7"
              data-testid="row"
            >
              <div
                class="c8"
              >
                <div
                  class="c11"
                >
                  <div
                    class="c12"
                    data-testid="avatar"
                    size="28"
                  >
                    <img
                      alt="Marta profile image"
                      class="c13"
                      src="https://images.pexels.com/photos/733872/pexels-photo-733872.jpeg?auto=compress&cs=tinysrgb&w=1260&h=750&dpr=2"
                    />
                  </div>
                  <p
                    class="c14"
                    font-weight="400"
                  >
                    Marta
                  </p>
                </div>
              </div>
              <div
                class="c8"
              >
                <div
                  class="c11"
                >
                  <p
                    class="c14"
                    font-weight="400"
                  >
                    Marta@gmail.com
                  </p>
                </div>
              </div>
              <div
                class="c8"
              >
                <div
                  class="c11"
                >
                  <p
                    class="c14"
                    font-weight="400"
                  >
                    Admin
                  </p>
                </div>
              </div>
              <div
                class="c8"
              >
                <div
                  class="c15"
                >
                  <button
                    class="c16"
                  >
                    <div
                      class="c17"
                      data-testid="icon-container"
                      size="24"
                    >
                      <span
                        class="c18 material-symbols-sharp"
                        color="#000000"
                        data-testid="icon"
                        size="24"
                      >
                        more_horiz
                      </span>
                    </div>
                  </button>
                </div>
              </div>
            </div>
            <div
              class="c7"
              data-testid="row"
            >
              <div
                class="c8"
              >
                <div
                  class="c11"
                >
                  <div
                    class="c12"
                    data-testid="avatar"
                    size="28"
                  >
                    <img
                      alt="Pedro profile image"
                      class="c13"
                      src="https://images.pexels.com/photos/428364/pexels-photo-428364.jpeg?auto=compress&cs=tinysrgb&w=1260&h=750&dpr=2"
                    />
                  </div>
                  <p
                    class="c14"
                    font-weight="400"
                  >
                    Pedro
                  </p>
                </div>
              </div>
              <div
                class="c8"
              >
                <div
                  class="c11"
                >
                  <p
                    class="c14"
                    font-weight="400"
                  >
                    pedro_dev@gmail.com
                  </p>
                </div>
              </div>
              <div
                class="c8"
              >
                <div
                  class="c11"
                >
                  <p
                    class="c14"
                    font-weight="400"
                  >
                    developer
                  </p>
                </div>
              </div>
              <div
                class="c8"
              >
                <div
                  class="c15"
                >
                  <button
                    class="c16"
                  >
                    <div
                      class="c17"
                      data-testid="icon-container"
                      size="24"
                    >
                      <span
                        class="c18 material-symbols-sharp"
                        color="#000000"
                        data-testid="icon"
                        size="24"
                      >
                        more_horiz
                      </span>
                    </div>
                  </button>
                </div>
              </div>
            </div>
            <div
              class="c7"
              data-testid="row"
            >
              <div
                class="c8"
              >
                <div
                  class="c11"
                >
                  <div
                    class="c12"
                    data-testid="avatar"
                    size="28"
                  >
                    <img
                      alt="Daniel profile image"
                      class="c13"
                      src="https://images.pexels.com/photos/220453/pexels-photo-220453.jpeg?auto=compress&cs=tinysrgb&w=1260&h=750&dpr=2"
                    />
                  </div>
                  <p
                    class="c14"
                    font-weight="400"
                  >
                    Daniel
                  </p>
                </div>
              </div>
              <div
                class="c8"
              >
                <div
                  class="c11"
                >
                  <p
                    class="c14"
                    font-weight="400"
                  >
                    daniel_ninja_dev@gmail.com
                  </p>
                </div>
              </div>
              <div
                class="c8"
              >
                <div
                  class="c11"
                >
                  <p
                    class="c14"
                    font-weight="400"
                  >
                    developer
                  </p>
                </div>
              </div>
              <div
                class="c8"
              >
                <div
                  class="c15"
                >
                  <button
                    class="c16"
                  >
                    <div
                      class="c17"
                      data-testid="icon-container"
                      size="24"
                    >
                      <span
                        class="c18 material-symbols-sharp"
                        color="#000000"
                        data-testid="icon"
                        size="24"
                      >
                        more_horiz
                      </span>
                    </div>
                  </button>
                </div>
              </div>
            </div>
            <div
              class="c7"
              data-testid="row"
            >
              <div
                class="c8"
              >
                <div
                  class="c11"
                >
                  <div
                    class="c12"
                    data-testid="avatar"
                    size="28"
                  >
                    <img
                      alt="Joana profile image"
                      class="c13"
                      src="https://images.pexels.com/photos/415829/pexels-photo-415829.jpeg?auto=compress&cs=tinysrgb&w=1260&h=750&dpr=2"
                    />
                  </div>
                  <p
                    class="c14"
                    font-weight="400"
                  >
                    Joana
                  </p>
                </div>
              </div>
              <div
                class="c8"
              >
                <div
                  class="c11"
                >
                  <p
                    class="c14"
                    font-weight="400"
                  >
                    joana_mac@gmail.com
                  </p>
                </div>
              </div>
              <div
                class="c8"
              >
                <div
                  class="c11"
                >
                  <p
                    class="c14"
                    font-weight="400"
                  >
                    QA
                  </p>
                </div>
              </div>
              <div
                class="c8"
              >
                <div
                  class="c15"
                >
                  <button
                    class="c16"
                  >
                    <div
                      class="c17"
                      data-testid="icon-container"
                      size="24"
                    >
                      <span
                        class="c18 material-symbols-sharp"
                        color="#000000"
                        data-testid="icon"
                        size="24"
                      >
                        more_horiz
                      </span>
                    </div>
                  </button>
                </div>
              </div>
            </div>
          </div>
        </div>
        <hr
          class="c19"
          data-testid="horizontal-rule"
        />
        <div
          class="c20"
        >
          <p
            class="c21"
            data-testid="setting-title"
          />
          <button
            class="c22"
          >
            Invite Member
          </button>
        </div>
      </div>
    </ul>
  </ul>
</DocumentFragment>
`;
